lib/onelogin/ruby-saml/logoutresponse.rb

```diff
@@ -69,46 +104,211 @@ def status_code
         end
       end
 
+      # @return [String] Gets the StatusMessage from a SAML Response.
+      #
+      def status_message
+        @status_message ||= begin
+          node = REXML::XPath.first(document, "/p:LogoutResponse/p:Status/p:StatusMessage", { "p" => PROTOCOL, "a" => ASSERTION })
+          node.text if node
+        end
+      end
+
+      # Gets the expected current_url
+      # (Right now we read this url from the Sinle Logout Service of the Settings)
+      # TODO: Calculate the real current_url and use it.
+      # @return [String] The current url
+      #
+      def current_url
+        @current_url ||= begin
+          if settings && settings.single_logout_service_url
+            settings.single_logout_service_url
+          end
+        end
+      end
+
+      # Validates the Logout Response (calls several validation methods)
+      # If fails, the attribute errors will contains the reason for the invalidation.
+      # @param soft [Boolean] soft Enable or Disable the soft mode (In order to raise exceptions when the logout response is invalid or not)
+      # @param request_id [String|nil] request_id The ID of the Logout Request sent by this SP to the IdP (if was sent any)
+      # @param get_params [Hash] In order to validate the In value we need to provide the GET parameters
+      # @return [Boolean] True if the Logout Response is valid, otherwise False if soft=True
+      # @raise [ValidationError] if soft == false and validation fails
+      #
+      def validate(soft = true, request_id = nil, get_params = nil)
```

Comment:
I think that the logic of this could be significantly cleaner.

If there was a boolean `@valid`(or another name with the same intent) instance variable on this class, you could drop all of the `&&` and simply mutate that variable to be returned on line 146. All of the validation methods are mutating ivars anyway, adding another would be a minimal change.